cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java

```diff
@@ -868,135 +875,370 @@ public Boolean call() throws Exception {
   }
 
   @Test
-  public void testCrossNSDatasetAccessWithAuthFlowlet() throws Exception {
+  public void testCrossNSFlowlet() throws Exception {
     createAuthNamespace();
-    getNamespaceAdmin().create(new NamespaceMeta.Builder()
-                                 .setName(CrossNsDatasetAccessApp.DATASET_OUTPUT_SPACE).build());
-    NamespaceId datasetOutputSpace = new NamespaceId(CrossNsDatasetAccessApp.DATASET_OUTPUT_SPACE);
-    addDatasetInstance(datasetOutputSpace.toId(), "keyValueTable", "store");
     ApplicationManager appManager = deployApplication(AUTH_NAMESPACE.toId(), CrossNsDatasetAccessApp.class);
 
-    // Send stream and run the flowlet as BOB. Should fail
+    // give BOB ALL permissions on the auth namespace so he can execute programs and also read the stream.
+    grantAndAssertSuccess(AUTH_NAMESPACE, BOB, EnumSet.of(Action.ALL));
+
+    // switch to BOB
+    SecurityRequestContext.setUserId(BOB.getName());
+
+    // Send data to stream as BOB this ensures that BOB can write to a stream in AUTH namespace
     StreamManager streamManager = getStreamManager(AUTH_NAMESPACE.toId(), CrossNsDatasetAccessApp.STREAM_NAME);
     for (int i = 0; i < 10; i++) {
       streamManager.send(String.valueOf(i).getBytes());
     }
+
+    // switch to back to ALICE
+    SecurityRequestContext.setUserId(ALICE.getName());
+
+    final FlowManager flowManager = appManager.getFlowManager(CrossNsDatasetAccessApp.FLOW_NAME);
+
+    testCrossNSSystemDatasetAccessWithAuthFlowlet(flowManager);
+    testCrossNSDatasetAccessWithAuthFlowlet(flowManager);
+
+    appManager.stopAll();
+  }
+
+  public void testCrossNSSystemDatasetAccessWithAuthFlowlet(final FlowManager flowManager) throws Exception {
+    addDatasetInstance(Id.Namespace.SYSTEM, "keyValueTable", "store");
+
+    // give bob write permission on the dataset
+    grantAndAssertSuccess(NamespaceId.SYSTEM.dataset("store"), BOB, EnumSet.of(Action.WRITE));
+
+    // switch to BOB
     SecurityRequestContext.setUserId(BOB.getName());
-    try {
-      FlowManager flowManager = appManager.getFlowManager(CrossNsDatasetAccessApp.FLOW_NAME);
-      flowManager.start();
-      Assert.fail("Bob shouldn't have the privlege to write to the dataset");
-    } catch (Exception e) {
-      // Expected
-    }
 
-    // Switch back to ALICE. Should Success
+    Map<String, String> args = ImmutableMap.of(
+      CrossNsDatasetAccessApp.OUTPUT_DATASET_NS, NamespaceId.SYSTEM.getNamespace(),
+      CrossNsDatasetAccessApp.OUTPUT_DATASET_NAME, "store"
+    );
+
+    // But trying to run a flow as BOB will fail since this flow writes to a dataset in system namespace
+    flowManager.start(args);
+    // wait for flow to be running
+    Tasks.waitFor(true, new Callable<Boolean>() {
+      @Override
+      public Boolean call() throws Exception {
+        return flowManager.isRunning();
```

Comment:
I am aware of waitForStatus(true) but according to my understanding it waits for started not running